File: minions/cluster/cl_minion.py
"""Minion for clusters."""
import gevent
import getopt
import os
import sys
import json
import uuid
import zerorpc

from minions import minion
import paramiko

from urlparse import urlparse

import pymongo
from pymongo import MongoClient

class ClusterMinion(minion.Minion):
    """Class definition."""

    def __init__(self, config):
        """Init function."""
        print("Initializing Cluster minion...")
        self._connected = False

        # Connect to DB (default "localhost")
        # db vars must be private to avoid zerorpc errors
        print("Connecting to DB...")
        try:
            self._db_client = MongoClient()
            self._db = self._db_client.test_db
        except Exception as e:
            print("Failed to get Storage database, reason: ", e)
            raise e

        try:
            self._db.images.create_index([
                ('id', pymongo.ASCENDING),
                ('name', pymongo.ASCENDING),
                ('minion', pymongo.ASCENDING)
            ], unique=True)
            self._db.sizes.create_index([
                ('id', pymongo.ASCENDING),
                ('name', pymongo.ASCENDING),
                ('minion', pymongo.ASCENDING)
            ], unique=True)
            self._db.instances.create_index([
                ('id', pymongo.ASCENDING),
                ('name', pymongo.ASCENDING),
                ('minion', pymongo.ASCENDING)
            ], unique=True)
        except Exception as e:
            print("Failed to create DB index, reason: ", e)
            raise e

        # Locks
        self._login_lock = False
        self._instance_lock = {}

        # Command to load bash environment in SSH
        self._cmd_env = ". /etc/profile; . ~/.bash_profile"

        # Login
        if config is not None:
            self.login(config)

        print("Initialization completed!")

    def getMinionName(self):
        return self.__class__.__name__

    def login(self, config):
        """Login to cluster using SSH."""

        ####################
        # Set the lock for login
        while self._login_lock:
            gevent.sleep(0)
        self._login_lock = True

        # Check if already connected
        if self._connected:
            print("Already connected...")
            self._login_lock = False
            return 0

        # Connect to provider
        self._config = config

        # Params
        if 'url' not in self._config:
            raise Exception(
                "Malformed login config. 'url' key not present."
            )
        if 'username' not in self._config:
            raise Exception(
                "Malformed login config. 'username' key not present."
            )
        if 'password' not in self._config:
            self._config['password'] = None

        # Get command line
        ssh = self._retrieveSSH()

        # Get configuration from cloud.json
        stdin, stdout, stderr = ssh.exec_command("cat cloud.json")
        json_config = stdout.read()
        try:
            config = json.loads(json_config)
            self._loadConfig(config)
        except Exception as e:
            print("Malformed config.json!")
            raise e

        # Close connection
        ssh.close()

        # Load instances
        instances = self._db.instances.find({
            "minion": self.getMinionName()
        })
        for inst in instances:
            # Init lock
            self._instance_lock[inst['id']] = False
            print("Loaded instance '{0}'".format(inst['id']))

        # Set connected
        self._connected = 1

        self._login_lock = False
        ####################

        return

    def createSize(self, size_cfg):
        """Create a new size and assign an UUID."""
        id = str(uuid.uuid1())
        size = {
            "_id": id,
            "id": id,
            "name": size_cfg['name'],
            "desc": "Description",
            "cpus": size_cfg['cpus'],
            "ram": size_cfg['ram'],
            "minion": self.__class__.__name__
        }
        self._db.sizes.insert_one(size)
        print("Created size:")
        print(size)
        return size['id']

    def createInstance(self, instance_cfg):
        """Reserve resources in cluster"""
        name = instance_cfg['name']

        # Get image
        image_id = instance_cfg['image_id']
        image = self._findImage(image_id)

        # Get size
        size_id = instance_cfg['size_id']
        size = self._findSize(size_id)

        print("=============================")
        print("Creating instance:")
        print("--> Name : {0}".format(name))
        print("--> Image: {0}".format(image))
        print("--> Size : {0}".format(size))

        # Save instance
        id = str(uuid.uuid1())
        instance = {
            '_id': id,
            'id': id,
            'name': name,
            'hostname': "rmoreno2@galgo.i3a.info",
            'image_id': image_id,
            'size_id': size_id,
            'exps': [],
            'workpath': image['workpath'],
            'inputpath': image['inputpath'],
            'minion': self.__class__.__name__,
            'ready': True
        }
        self._instance_lock[id] = False
        self._db.instances.insert_one(instance)
        print('Instance "{0}" (DUMMY) created'.format(instance['id']))

        return instance['id']

    def destroyInstance(self, instance_id):
        """Destroy an existing instance"""

        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance
        instance = self._findInstance(instance_id)

        # Get instance command line
        ssh = self._retrieveSSH()
        if ssh is None:
            self._instance_lock[instance_id] = False
            raise Exception("Instance without SSH")

        # Delete job
        if 'job_id' in instance:
            cmd = 'qdel -W force {0}'.format(instance['job_id'])
            task = gevent.spawn(self._executeSSH, ssh, cmd)
            gevent.joinall([task])

        # Close SSH
        ssh.close()

        # Remove from DB
        self._db.instances.remove(
            {'id': instance_id, 'minion': self.getMinionName()}
        )

        # Remove lock
        self._instance_lock.pop(instance_id, None)
        ####################

    def getImages(self, filter=None):
        """Get image list using an optional filter."""
        if filter is None:
            return list(
                self._db.images.find({'minion': self.__class__.__name__})
            )
        else:
            image = self._db.images.find_one({
                'minion': self.__class__.__name__,
                'id': filter
            })
            if image is None:
                return list(self._db.images.find({
                    'minion': self.__class__.__name__,
                    'name': {'$regex': '.*' + filter + '.*'}
                }))
            else:
                return image

    def getSizes(self, filter=None):
        """Get size list using an optional name filter."""
        if filter is None:
            return list(
                self._db.sizes.find({'minion': self.__class__.__name__})
            )
        else:
            size = self._db.sizes.find_one({
                'minion': self.__class__.__name__,
                'id': filter
            })
            if size is None:
                return list(self._db.sizes.find({
                    'minion': self.__class__.__name__,
                    'name': {'$regex': '.*' + filter + '.*'}
                }))
            else:
                return size

    def getInstances(self, filter=None):
        """Get instance list using an optional name filter."""
        if filter is None:
            return list(
                self._db.instances.find({'minion': self.__class__.__name__})
            )
        else:
            inst = self._db.instances.find_one({
                'minion': self.__class__.__name__,
                'id': filter
            })
            if inst is None:
                return list(self._db.instances.find({
                    'minion': self.__class__.__name__,
                    'name': {'$regex': '.*' + filter + '.*'}
                }))
            else:
                return inst

    def getInstanceHostname(self, instance_id):
        """Get instance hostname"""

        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance
        instance = self._findInstance(instance_id)

        # Check if instance is ready
        if not instance['ready']:
            self._instance_lock[instance_id] = False
            raise Exception("Instance {0} is not ready, unable to retrieve hostname".format(
                instance['id']
            ))

        self._instance_lock[instance_id] = False
        ####################

        return "rmoreno2@galgo.i3a.info"

    def executeCommand(self, cmd, instance_id):
        # TODO: Check instance minion

        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance
        instance = self._findInstance(instance_id)

        # Check if instance is ready
        if not instance['ready']:
            self._instance_lock[instance_id] = False
            raise Exception("Instance {0} is not ready, unable to execute command: \n'{1}'".format(
                instance['id'], cmd
            ))

        # Get instance command line
        ssh = self._retrieveSSH()
        if ssh is None:
            self._instance_lock[instance_id] = False
            raise Exception("Instance without SSH")

        # Execute task
        cmd = """{0}; {1}""".format(self._cmd_env, cmd)
        task = gevent.spawn(self._executeSSH, ssh, cmd)
        gevent.joinall([task])
        ret_val = task.value[0].read()
        ret_err = task.value[1].read()

        # Close connection
        ssh.close()

        self._instance_lock[instance_id] = False
        ####################

        return {
            'stdout': ret_val,
            'stderr': ret_err
        }

    def executeScript(self, script, work_dir, instance_id, nodes):
        # TODO: Check instance minion

        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance
        instance = self._findInstance(instance_id)

        # Check if instance is ready
        if not instance['ready']:
            self._instance_lock[instance_id] = False
            raise Exception("Instance {0} is not ready, unable to execute script: \n'{1}'".format(
                instance['id'], script
            ))

        # Get instance command line
        ssh = self._retrieveSSH()
        if ssh is None:
            self._instance_lock[instance_id] = False
            raise Exception("Instance without SSH")

        # Get size
        size = self._findSize(instance['size_id'])

        # QSUB launch command
        qsub_cmd = "qsub -N {0}-{1}-{2} -l select={0}:ncpus={1}:mem={2}MB -o {3} -e {3}".format(
            nodes, size['cpus'], size['ram'], work_dir
        )
        cmd = """{0}; echo '{1}' | {2} """.format(
            self._cmd_env, script, qsub_cmd
        )

        # Execute script
        task = gevent.spawn(self._executeSSH, ssh, cmd)
        gevent.joinall([task])
        job_id = task.value[0].read()

        # Close connection
        ssh.close()

        self._instance_lock[instance_id] = False
        ####################

        return job_id

    def cleanJob(self, job_id, instance_id):
        """Delete job from queue system"""

        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance command line
        ssh = self._retrieveSSH()
        if ssh is None:
            self._instance_lock[instance_id] = False
            raise Exception("Instance without SSH")

        # Check if experiment is in jobs queue and terminate
        if job_id is not None:
            print("Cleaning job ID: {0}".format(job_id))
            while True:
                cmd = """{0}; qdel -W force {1}""".format(
                    self._cmd_env, job_id
                )
                task = gevent.spawn(self._executeSSH, ssh, cmd)
                gevent.joinall([task])
                ret = task.value[1].read()
                gevent.sleep(1)
                if ret != "":
                    break

        # Close connection
        ssh.close()

        self._instance_lock[instance_id] = False
        ####################

    def getJobStatus(self, job_id, instance_id):
        ####################
        # Set the lock for the instance
        while self._instance_lock[instance_id]:
            gevent.sleep(0)
        self._instance_lock[instance_id] = True

        # Get instance command line
        ssh = self._retrieveSSH()
        if ssh is None:
            self._instance_lock[instance_id] = False
            raise Exception("Instance without SSH")

        # Check if experiment is in jobs queue and terminate
        status = "unknown"
        if job_id is not None:
            cmd = """{0}; qstat {1}""".format(
                self._cmd_env, job_id
            )
            task = gevent.spawn(self._executeSSH, ssh, cmd)
            gevent.joinall([task])
            ret = task.value[1].read()
            if "Unknown" not in ret:
                status = "running"
            else:
                status = "finished"

        # Close connection
        ssh.close()

        self._instance_lock[instance_id] = False
        ####################

        return status

    """ Private functions """
    def _loadConfig(self, config):
        print("Loading config...")

        # Get existing data from DB
        prev_images = self._db.images.find({'minion': self.__class__.__name__})
        prev_sizes = self._db.sizes.find({'minion': self.__class__.__name__})

        # Parse images
        print("Parsing images...")
        for image in config['images']:
            # Check if already exists
            found = False
            for prev_img in prev_images:
                if image['name'] == prev_img['name']:
                    # Found!
                    found = True
                    print('-- Image "{0}" already exists'.format(image['name']))
                    break
            if not found:
                id = str(uuid.uuid1())
                image['_id'] = id
                image['id'] = id
                image['minion'] = self.__class__.__name__
                print("-- Adding image: {0}".format(image))
                self._db.images.insert_one(image)

        # Parse sizes
        print("Parsing sizes...")
        for size in config['sizes']:
            # Check if already exists
            found = False
            for prev_size in prev_sizes:
                if size['name'] == prev_size['name'] \
                    and size['cpus'] == prev_size['cpus'] \
                        and size['ram'] == prev_size['ram']:
                    # Found!
                    found = True
                    print('-- Size "{0}" already exists'.format(size['name']))
                    break
            if not found:
                id = str(uuid.uuid1())
                size['_id'] = id
                size['id'] = id
                size['minion'] = self.__class__.__name__
                print("-- Adding size: {0}".format(size))
                self._db.sizes.insert_one(size)

        print("Config loaded!")

    def _findImage(self, image_id):
        """Return image data"""
        return self._db.images.find_one({
            'minion': self.__class__.__name__,
            'id': image_id
        })

    def _findSize(self, size_id):
        """Return size data"""
        return self._db.sizes.find_one({
            'minion': self.__class__.__name__,
            'id': size_id
        })

    def _findInstance(self, inst_id):
        """Return size data"""
        return self._db.instances.find_one({
            'minion': self.__class__.__name__,
            'id': inst_id
        })

    def _retrieveSSH(self):
        # Get config
        url = self._config['url']
        username = self._config['username']
        # password = self._config['password']

        # Get valid URL
        url = urlparse(url).path
        try:
            # Create command line and connect to the cluster
            ssh = paramiko.SSHClient()
            ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            ssh.connect(url, username=username)

        except Exception as e:
            print("FAILED to connect to", url, "- Reason:", e)
            raise e

        # Save connection var
        return ssh

    def _executeSSH(self, ssh, cmd):
        stdin, stdout, stderr = ssh.exec_command(cmd)
        while not stdout.channel.exit_status_ready():
            gevent.sleep(0)
        return [stdout, stderr]

# Start RPC minion
# Execute this only if called directly from python command
# From now RPC is waiting for requests
if __name__ == "__main__":
    # Get arguments
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hc:", ["cfg="])
        if len(opts) == 0:
            raise getopt.GetoptError("Please, specify configuration file with -c")
    except getopt.GetoptError as e:
        # Not valid arguments
        print(e)
        print("Usage: python -m minions.cluster.cl_minion -c <config_file>")
        sys.exit(2)

    # Iterate arguments
    for opt, arg in opts:
        if opt == '-h':
            print("python -m minion.cluster.cl_minion -c <config_file>")
            sys.exit()
        elif opt in ("-c", "--cfg"):
            cfg_file = arg

    print("Config file: {0}".format(cfg_file))

    # Check if config file exists
    if os.path.isfile(cfg_file):
        with open(cfg_file) as f:
            cfg = json.load(f)
    else:
        print("Inexistent file: {0}".format(cfg_file))
        sys.exit(3)

    # Read configuration file
    rpc = zerorpc.Server(ClusterMinion(cfg), heartbeat=30)
    rpc.bind("tcp://0.0.0.0:8238")
    rpc.run()

File: minions/cluster/cl_minion.js
/**
 * Copyright 2017 University of Castilla - La Mancha
 */
var zerorpc = require('zerorpc');
var async = require('async');
var request = require('request');
var fs = require('fs');
var exec = require('child_process').exec;
var sleep = require('sleep');
var ssh2 = require('ssh2').Client;
var mongo = require('mongodb').MongoClient;
var utils = require('../../overlord/utils');
var logger = utils.logger;

/***********************************************************
 * --------------------------------------------------------
 * MINION NAME
 * --------------------------------------------------------
 ***********************************************************/
var MINION_NAME = "ClusterMinion";

/***********************************************************
 * --------------------------------------------------------
 * GLOBAL VARS
 * --------------------------------------------------------
 ***********************************************************/
var zserver = null;
var cfg = process.argv[2];
var constants = {};

var ssh_conn = null;
var cmd_env = ". /etc/profile; . ~/.bash_profile";

/***********************************************************
 * --------------------------------------------------------
 * METHODS
 * --------------------------------------------------------
 ***********************************************************/

/**
 * Login to provider.
 */
var login = function(cb){
   // Check if already connected
   if(ssh_conn){
      logger.info('['+MINION_NAME+'] Already connected...');
      return cb(null);
   }

   var private_key = null;
   if(constants.authkeyfile) private_key = fs.readFileSync(constants.authkeyfile);

   // Create connection
   utils.connectSSH(constants.username, constants.url, private_key, 30000, function(error, conn){
      if(error) return cb(error);

      // Save connection
      ssh_conn = conn;
      return cb(null);
   });
}

var getImages = function(filter, getCallback){
   if(!database) return getCallback(new Error("No connection to DB."));

   // Retrieve images
   var query = {minion: MINION_NAME};
   if(filter) query.id = filter;
   database.collection('images').find(query).toArray().then(
      function(images){
         // Callback
         if(filter && images.length == 0) return getCallback(new Error("Image '"+filter+"' not found."));

         // Update quotas
         getQuotas(function(error, quotas){
            if(error) return getCallback(error);

            // Set quotas
            for(var i = 0; i < images.length; i++){
               images[i].quotas = quotas;
            }

            if(filter && images.length == 1) return getCallback(null, images[0]);
            getCallback(null, images);
         });
      }
   );
}

var getSizes = function(filter, getCallback){
   if(!database) return getCallback(new Error("No connection to DB."));

   // Retrieve sizes
   var query = {minion: MINION_NAME};
   if(filter) query.id = filter;
   database.collection('sizes').find(query).toArray().then(
      function(sizes){
         // Callback
         if(filter && sizes.length == 0) return getCallback(new Error("Size '"+filter+"' not found."));
         if(filter && sizes.length == 1) return getCallback(null, sizes[0]);
         getCallback(null, sizes);
      }
   );
}

var getInstances = function(filter, getCallback){
   if(!database) return getCallback(new Error("No connection to DB."));

   // Retrieve instances
   var query = {minion: MINION_NAME};
   if(filter) query.id = filter;
   database.collection('instances').find(query).toArray().then(
      function(instances){
         // Callback
         if(filter && instances.length == 0) return getCallback(new Error("Instance '"+filter+"' not found."));
         if(filter && instances.length == 1) return getCallback(null, instances[0]);
         getCallback(null, instances);
      }
   );
}

var createInstance = function(inst_cfg, createCallback){
   // Check params
   if(!inst_cfg) return createCallback(new Error("Null inst_cfg."));
   if(!inst_cfg.name) return createCallback(new Error("No name in inst_cfg."));
   if(!inst_cfg.image_id) return createCallback(new Error("No image_id in inst_cfg."));
   if(!inst_cfg.size_id) return createCallback(new Error("No size_id in inst_cfg."));
   if(!inst_cfg.nodes) return createCallback(new Error("No nodes in inst_cfg."));

   // Retrieve image
   getImages(inst_cfg.image_id, function(error, image){
      if(error) return createCallback(error);
      getSizes(inst_cfg.size_id, function(error, size){
         if(error) return createCallback(error);

         // Create instance metadata
         var id = utils.generateUUID();
         var inst = {
            _id: id,
            id: id,
            name: inst_cfg.name,
            image_id: inst_cfg.image_id,
            size_id: inst_cfg.size_id,
            execs: [],
            nodes: inst_cfg.nodes,
            size: {
               cpus: size['cpus'],
               ram: size['ram']
            },
            image: {
               workpath: image['workpath'],
               inputpath: image['inputpath'],
               outputpath: image['outputpath'],
               libpath: image['libpath'],
               tmppath: image['tmppath']
            },
            minion: MINION_NAME,
            hostname: constants.url,
            ip: constants.url,
            members: null,
            in_use: true,
            idle_time: Date.now(),
            ready: true
         };

         // Add to DB
         database.collection('instances').insert(inst, function(error){
            if(error) return createCallback(error);
            logger.log('['+MINION_NAME+']['+inst.id+'] Added instance to DB.');
            createCallback(null, inst.id);
         });
      });
   });
}

var destroyInstance = function(inst_id, destroyCallback){
   // Remove instance from DB
   database.collection('instances').remove({_id: inst_id});
   return destroyCallback(null);
}

var executeScript = function(inst_id, script, work_dir, nodes, executeCallback){
   // Available connection?
   if(!ssh_conn) return wfcb('No SSH connection is present');

   // Always have work_dir
   if(!work_dir) work_dir = "~";

   async.waterfall([
      // Get instance
      function(wfcb){
         getInstances(inst_id, wfcb);
      },
      // Execute script with queue system
      function(inst, wfcb){
         // Check if instance is ready
         if(inst.ready != true) return wfcb(new Error('Instance "'+inst_id+'" is not ready, unable to execute script.'));

         // QSUB launch command
         var qsub_cmd = "qsub -N "+inst.nodes+"-"+inst.size.cpus+"-"+inst.size.ram+" -l select="+inst.nodes+":ncpus="+inst.size.cpus+":mem="+inst.size.ram+"MB -e "+inst.image.tmppath+"/${PBS_JOBID}.stderr -o "+inst.image.tmppath+"/${PBS_JOBID}.stdout";
         var cmd = cmd_env+"; echo '"+script+"; echo -n $? > "+inst.image.tmppath+"/${PBS_JOBID}.code' | "+qsub_cmd+" | tr -d '\n'"; // Strip new line

         logger.debug("["+MINION_NAME+"]["+inst_id+"] Executing script: " + cmd);
         // Execute (blocking as we are using queue system)
         utils.execSSH(ssh_conn, cmd, work_dir, true, null, wfcb);
      }
   ],
   function(error, output){
      return executeCallback(error, output.stdout); // Return Job ID
   });
}

var executeCommand = function(inst_id, script, executeCallback){
   // Available connection?
   if(!ssh_conn) return wfcb('No SSH connection is present');

   async.waterfall([
      // Get instance
      function(wfcb){
         getInstances(inst_id, wfcb);
      },
      // Execute command
      function(inst, wfcb){
         // Check if instance is ready
         if(inst.ready != true) return wfcb(new Error('Instance "'+inst_id+'" is not ready, unable to execute script.'));

         // Just the command
         var cmd = cmd_env+"; "+script;

         // Execute
         logger.debug("["+MINION_NAME+"]["+inst_id+"] Executing command: " + cmd);
         utils.execSSH(ssh_conn, cmd, null, true, null, wfcb);
      }
   ],
   function(error, output){
      return executeCallback(error, output);
   });
}

var getJobStatus = function(job_id, inst_id, getCallback){
   // Available connection?
   if(!ssh_conn) return getCallback('No SSH connection is present');

   var status = "unknown";
   if(job_id){
      var cmd = cmd_env+"; qstat "+job_id;
      utils.execSSH(ssh_conn, cmd, null, true, null, function(error, output){
         // Parse output to get job info
         if(output.stderr.indexOf("Unknown") == -1){
            status = "running";
         } else {
            status = "finished";
         }
         return getCallback(null, status);
      });
   } else {
      return getCallback(null, status);
   }
}

var getQuotas = function(getCallback){
   // Return quotas
   var quotas = {
      cores: {
         in_use: 0,
         limit: 1000
      },
      floating_ips: {
         in_use: 0,
         limit: 1000
      },
      instances: {
         in_use: 0,
         limit: 1000
      },
      ram: {
         in_use: 0,
         limit: 1000000
      }
   };
   return getCallback(null, quotas);
}

var cleanJob = function(job_id, inst_id, cleanCallback){
   // QDEL
   var cmd = cmd_env+'; qdel -W force '+job_id;
   utils.execSSH(ssh_conn, cmd, null, true, null, function(error, output){
      return cleanCallback(error);
   })
}

var _loadConfig = function(config, loadCallback){
   if(!config.db) return loadCallback(new Error("No db field in CFG."));
   if(!config.listen) return loadCallback(new Error("No listen field in CFG."));
   if(!config.url) return loadCallback(new Error("No url field in CFG."));
   if(!config.username) return loadCallback(new Error("No username field in CFG."));

   async.waterfall([
      // Connect to MongoDB
      function(wfcb){
         // Connect to database
         logger.info("["+MINION_NAME+"] Connecting to MongoDB: " + config.db);
         mongo.connect(config.db, function(error, db){
            if(error) throw error;
            logger.info("["+MINION_NAME+"] Successfull connection to DB");

            // Set global var
            database = db;

            // Next
            return wfcb(null);
         });
      },
      // Connect to cluster
      function(wfcb){
         logger.info("["+MINION_NAME+"] Connecting to '"+config.username+"@"+config.url+"'...");
         login(function(error){
            if(error) return wfcb(error);

	    // Test connection
            //utils.execSSH(ssh_conn, "ls", "$HOME", true, function(error, output){
            //   logger.info("["+MINION_NAME+"] ls test: "+ error);
	    //});

            logger.info("["+MINION_NAME+"] Successfull connection to cluster.");

            // Next
            return wfcb(null);
         });
      },
      // Setup RPC
      function(wfcb){
         // Setup server
         zserver = new zerorpc.Server({
            login: login,
            autoupdate: autoupdate,
            getMinionName: getMinionName,
            getImages: getImages,
            getSizes: getSizes,
            getInstances: getInstances,
            createInstance: createInstance,
            destroyInstance: destroyInstance,
            executeScript: executeScript,
            executeCommand: executeCommand,
            getJobStatus: getJobStatus,
            cleanJob: cleanJob,
            getQuotas: getQuotas
         }, 30000);

         // Listen
         logger.info("["+MINION_NAME+"] Listening on "+config.listen);
         zserver.bind(config.listen);

         // RPC error handling
         zserver.on("error", function(error){
            logger.error("["+MINION_NAME+"] RPC server error: "+ error);
         });

         // Next
         wfcb(null);
      },
      // Load images
      function(wfcb){
         logger.info("["+MINION_NAME+"] Loading images...");
         // Get images from DB
         database.collection('images').find({minion: MINION_NAME}).toArray().then(function(images){
            for(var i = 0; i < config.images.length; i++){
               var image = config.images[i];

               // Search image
               var found = false;
               for(var j = 0; j < images.length; j++){
                  if(image.internal_id == images[j].internal_id){
                     // Found
                     logger.info('['+MINION_NAME+'] Image "'+image.internal_id+'" exists already.');
                     found = true;
                     break;
                  }
               }

               // Add to DB
               if(!found){
                  image._id = utils.generateUUID();
                  image.id = image._id;
                  image.minion = MINION_NAME;
                  database.collection('images').insert(image);
                  logger.info('['+MINION_NAME+'] Added image "'+image.id+'" to database.')
               }
            }

            // Next
            wfcb(null);
         });
      },
      // Load sizes
      function(wfcb){
         logger.info("["+MINION_NAME+"] Loading sizes...");
         // Get sizes from DB
         database.collection('sizes').find({minion: MINION_NAME}).toArray().then(function(sizes){
            for(var i = 0; i < config.sizes.length; i++){
               var size = config.sizes[i];

               // Search size
               var found = false;
               for(var j = 0; j < sizes.length; j++){
                  if(size.internal_id == sizes[j].internal_id){
                     // Found
                     logger.info('['+MINION_NAME+'] Size "'+size.internal_id+'" exists already.');
                     found = true;
                     break;
                  }
               }

               // Add to DB
               if(!found){
                  size._id = utils.generateUUID();
                  size.id = size._id;
                  size.minion = MINION_NAME;
                  database.collection('sizes').insert(size);
                  logger.info('['+MINION_NAME+'] Added size "'+size.id+'" to database.')
               }
            }

            // Next
            wfcb(null);
         });
      },
      // Setup compatibility
      function(wfcb){
         logger.info("["+MINION_NAME+"] Setting compatibility between images and sizes...");
         // Get images from DB
         database.collection('images').find({minion: MINION_NAME}).toArray().then(function(images){
            // Get sizes from DB
            database.collection('sizes').find({minion: MINION_NAME}).toArray().then(function(sizes){
               // Iterate images
               for(var i = 0; i < images.length; i++){
                  // Sizes for this image
                  var sizes_compatible = [];
                  var image = images[i];
                  // Iterate sizes
                  for(var j = 0; j < sizes.length; j++){
                     var size = sizes[j];

                     // Compatibles?
                     for(var z = 0; z < image.internal_sizes_compatible.length; z++){
                        var comp_id = image.internal_sizes_compatible[z];
                        if(size.internal_id == comp_id){
                           // Compatible!
                           sizes_compatible.push(size.id);
                        }
                     }
                  }
                  // Save changes
                  logger.info('['+MINION_NAME+'] Sizes "'+sizes_compatible+'" are compatible with image "'+image.id+'".');
                  database.collection('images').updateOne({id: image.id},{$set:{sizes_compatible:sizes_compatible}});
               }

               // Next
               return wfcb(null);
            });
         });
      },
   ],
   function(error){
      if(error){
         loadCallback(error);
      } else {
         logger.info("["+MINION_NAME+"] Config loaded.");
         loadCallback(null);
      }
   });
}

/**
 * Update repository and restart process.
 */
var autoupdate = function(cb){
   logger.info('['+MODULE_NAME+'] Autoupdate: Begin.');
   async.waterfall([
      // Check configuration
      function(wfcb){
         if(!constants.AUTOUPDATE){
            return wfcb(new Error('Autoupdate is not enabled.'));
         }
         return wfcb(null);
      },
      // Update repository
      function(wfcb){
         exec('git pull origin '+constants.AUTOUPDATE, function(error, stdout, stderr){
            return wfcb(error);
         });
      }
   ],
   function(error){
      if(error){
         // Error trying to checkpoint execution
         logger.debug('['+MODULE_NAME+'] Autoupdate: Error.');
         cb(error);
      } else {
         logger.debug('['+MODULE_NAME+'] Autoupdate: Done, resetting...');
         // Callback before exit
         cb(null);
         // Exit from Node, forever will restart the service.
         process.exit(1);
      }
   });
}

var getMinionName = function(getCallback){
   getCallback(null, MINION_NAME);
}

/***********************************************************
 * --------------------------------------------------------
 * INITIALIZATION
 * --------------------------------------------------------
 ***********************************************************/

// Get config file
if(!cfg) throw new Error("No CFG file has been provided.");

// Steps
async.waterfall([
   // Read config file
   function(wfcb){
      logger.info("["+MINION_NAME+"] Reading config file: "+cfg);
      fs.readFile(cfg, function(error, fcontent){
         if(error) return wfcb(error);
         return wfcb(null, fcontent);
      });
   },
   // Load cfg
   function(fcontent, wfcb){
      logger.info("["+MINION_NAME+"] Loading config file...");

      // Parse cfg
      constants = JSON.parse(fcontent);

      // Loading
      _loadConfig(constants, function(error){
         if(error) return wfcb(error);
         return wfcb(null);
      });
   }
],
function(error){
   if(error) throw error;
   logger.info('['+MINION_NAME+'] Initialization completed.');
   //__test();
});

